### PR TITLE
Add VSL rate limiting

### PR DESCRIPTION
**Pull request by daghf**
This adds rate limiting to varnishncsa and varnishlog.

Rate limiting is done on a per-transaction basis, respective to the grouping mode selected. I.e. for -g request the limit will apply on a per-request basis, -g raw on a per-record basis, etc.

Rate limit is specified as ``-R <limit>[/period]``, where period can be one of 's', 'm', 'h' or 'd'. Default period if not specified is seconds ('s').

This is particularly useful for setups that want to run varnishlog or varnishncsa logging to disk looking for particular error situations, without being concerned about running out of disk space in a hurry.

The syntax is inspired by [netfilter](https://netfilter.org/documentation/HOWTO/packet-filtering-HOWTO-7.html]) (search for _limit_)

### CODE REVIEW

File: lib/libvarnishapi/vsl_dispatch.c [hunk 1]
```diff
@@ -908,10 +912,32 @@ vtx_force(struct VSLQ *vslq, struct vtx *vtx, const char *reason)
 	AN(vtx->flags & VTX_F_COMPLETE);
 }
 
+static int
+vslq_ratelimit(struct VSLQ *vslq)
+{
+	double now, delta;
```
Comment (nigoroll) [inline lib/libvarnishapi/vsl_dispatch.c]: vtim_mono please

File: lib/libvarnishapi/vsl_dispatch.c [hunk 2]
```diff
@@ -194,6 +194,10 @@ struct VSLQ {
 	VTAILQ_HEAD(,vtx)	cache;
 	unsigned		n_cache;
 
+	/* Rate limiting */
+	double			credits;
```
Comment (nigoroll) [inline lib/libvarnishapi/vsl_dispatch.c]: vtim_mono please

File: lib/libvarnishapi/vsl_api.h [hunk 3]
```diff
@@ -84,6 +84,8 @@ struct VSL_data {
 	int				c_opt;
 	int				C_opt;
 	int				L_opt;
+	int				R_opt_l;
+	double				R_opt_p;
```
Comment (nigoroll) [inline lib/libvarnishapi/vsl_api.h]: vtim_dur

File: lib/libvarnishapi/vsl_arg.c [hunk 4]
```diff
@@ -300,6 +300,45 @@ vsl_IX_arg(struct VSL_data *vsl, int opt, const char *arg)
...
+	AN(p);
+	if (*p == '\0') {
+		vsl->R_opt_p = 1.0;
+		return (1);
+	}
+
+	if (*p != '/' || p[1] == '\0')
+		return (vsl_diag(vsl, "-R: Syntax error"));
+	p++;
+	if (!strcmp(p, "s"))
```
Comment (nigoroll) [inline lib/libvarnishapi/vsl_arg.c]: use VNUM_duration ?

**Comment by daghf** [comment]
Merged via b2b04a617d83a834e8739ddde897ef7211945c8a

**Pull request closed without merging**